### PR TITLE
fix upload error when cookbook tarball uid/gid is very large

### LINKED ISSUE #1806: supermarket upload errors when cookbook contents is owned by very large uid or gid
Copied from https://github.com/chef/chef-dk/issues/2018 as I'm also seeing the same issue.

# Description
When trying to upload cookbook to Chef Supermarket, I get the error `ERROR: Multipart POST part 'tarball' is corrupt: "\x80\x00\x00\x00\x0E:\xBFD" is not an octal string`
 
Example:
 ```
$ knife supermarket share stig -VV -o ../ -c ~/.chef/knife-chefio.rb
INFO: Using configuration from /Users/isuftin/.chef/knife-chefio.rb
INFO: Validating ruby files
INFO: Validating templates
INFO: Syntax OK
Generating metadata for stig from /var/folders/58/n9bfwg050sdg1dmn9688wrcm001rp_/T/chef-stig-build20190410-52953-ff735p/stig/metadata.rb
Making tarball stig.tgz
DEBUG: Signing: method: post, url: https://supermarket.chef.io/api/v1/cookbooks, file: #<File:0x00007ff745d18988>, User-id: isuftin-usgs, Timestamp: 2019-04-10T13:06:07Z
ERROR: Multipart POST part 'tarball' is corrupt: "\x80\x00\x00\x00\x0E:\xBFD" is not an octal string
```
I did try switching gem versions as mentioned here:

* [#1549 (comment)](https://github.com/chef/chef-dk/issues/1549#issuecomment-387342439)

This did not help.

Possible related issues:

* [rubygems/rubygems#2213](https://github.com/rubygems/rubygems/issues/2213)
* [cerner/cerner_splunk#143](https://github.com/cerner/cerner_splunk/issues/143)

# ChefDK Version
Using Chef Workstation.
```
$ chef -v
Chef Workstation: 0.2.53
  chef-run: 0.2.8
  chef-client: 14.10.9
  delivery-cli: 0.0.52 (9d07501a3b347cc687c902319d23dc32dd5fa621)
  berks: 7.0.7
  test-kitchen: 1.24.0
  inspec: 3.6.6
```
```
$ chef gem -v
2.7.6
```
# Platform Version
MacOS:
```
$ sw_vers
ProductName:    Mac OS X
ProductVersion: 10.12.6
BuildVersion:   16G1917
```
# Replication Case
What i do locally for a cloned tagged checkout of https://github.com/USGS-CIDA/stig:
```
$ knife cookbook metadata stig -o ../
Generating metadata for stig from /Users/isuftin/Chef/cookbooks/stig/metadata.rb

$ knife supermarket share stig -VV -o ../ -c ~/.chef/knife-chefio.rb

INFO: Using configuration from /Users/isuftin/.chef/knife-chefio.rb
INFO: Validating ruby files
INFO: Validating templates
INFO: Syntax OK
Generating metadata for stig from /var/folders/58/n9bfwg050sdg1dmn9688wrcm001rp_/T/chef-stig-build20190410-60032-1e5qcb1/stig/metadata.rb
Making tarball stig.tgz
DEBUG: Signing: method: post, url: https://supermarket.chef.io/api/v1/cookbooks, file: #<File:0x00007fc414f2f398>, User-id: isuftin-usgs, Timestamp: 2019-04-10T13:14:14Z
ERROR: Multipart POST part 'tarball' is corrupt: "\x80\x00\x00\x00\x0E:\xBFD" is not an octal string
```
My knife-chefio.rb file looks like:
```
$ cat  ~/.chef/knife-chefio.rb
current_dir = File.dirname(__FILE__)
log_level                :info
log_location             STDOUT
node_name                "isuftin-usgs"
client_key               "#{current_dir}/isuftin-usgs-chefio.pem"
chef_server_url          "https://supermarket.chef.io/"
```
# Stacktrace
https://gist.github.com/isuftin/e8bb0804d5510a0258ead2ea65638fca

### PR DESCRIPTION
**Pull request by robbkidd**
## Description

RubyGems' Gem::Package::TarReader has some strict octal checking that is incompatible with pax format tar files with large (greater than 8^8) uid or gid entries (see rubygems/rubygems#2213). This can happen particularly when publishing a cookbook on a Windows workstation that is a member of a domain; the group IDs there can be very large numbers.

This change swaps out the handling of tar files from RubyGems' library to Chef's own ffi-libarchive. The previous logic also assumed that cookbook tarballs must be GZipped which is probably an assumption made elsewhere in the Chef ecosystem. Because libarchive is pretty forgiving of the handling of multiple formats of tar files, the requirement for GZipped tar files is preserved and implemented with FileMagic, already a dependency of Supermarket via the Fieri engine subcomponent.

## Related Issue

Fixes #1806

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- ~New feature (non-breaking change which adds functionality)~
- ~Breaking change (fix or feature that would cause existing functionality to change)~
- ~Chore (non-breaking change that does not add functionality or fix an issue)~

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- ~I have updated the documentation accordingly.~
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by robbkidd** [comment]
@teknofire Look past all the force pushing and this is where I would up after our pairing on Friday. Thoughts?

File: src/supermarket/app/models/cookbook_upload/archive.rb [hunk 1]
```diff
@@ -78,26 +81,41 @@ def read(path)
...
       end
     end
+
+    #
+    # Determines whether a file at a path is gzipped or not
+    #
+    # @return true/false
+    #
+    def gzipped?
+      FileMagic.mime.file(@source.path).match? %r{application\/x?-?gzip;}
```
Comment (tyler-ball) [inline src/supermarket/app/models/cookbook_upload/archive.rb]: What is an example of the mime string that will be returned? I'm not familiar with FileMagic but just trying to ensure this regex will match

File: src/supermarket/app/models/cookbook_upload/archive.rb [hunk 2]
```diff
@@ -78,26 +81,41 @@ def read(path)
...
+        ::Archive::Reader.open_filename(@source.path) do |tar|
+          tar.each_entry_with_data do |entry, data|
+            contents = data.is_a?(String) ? data : ''
+            yield entry, contents
           end
         end
-      rescue Zlib::GzipFile::Error => e
-        raise Error, e.message
+      rescue ::Archive::Error => e
+        case e.message
```
Comment (teknofire) [inline src/supermarket/app/models/cookbook_upload/archive.rb]: Should add an `else` here to reraise the error if it comes back with a different message so we don't silently swallow it.

File: src/supermarket/app/models/cookbook_upload/archive.rb (hunk 1 above)
Comment (robbkidd) [reply src/supermarket/app/models/cookbook_upload/archive.rb]: I saw `"application/x-gzip; charset=binary"` on macOS and  then `"application/gzip; charset=binary"` on Linux in Travis. (The reason for several of the force pushes.) I used [Rubular to tinker with the regex](https://rubular.com/r/bNfFNbNYqqFof4).

File: src/supermarket/app/models/cookbook_upload/archive.rb (hunk 2 above)
Comment (robbkidd) [reply src/supermarket/app/models/cookbook_upload/archive.rb]: Good catch. That's very true. Just say no to error swallowing.

Comment (robbkidd) [reply src/supermarket/app/models/cookbook_upload/archive.rb]: [Added an else](https://github.com/chef/supermarket/pull/1810/commits/4507051cc8bf972fe817462c81652a6d8128de7c#diff-b112e427d9f10480a410ca9caf8081ccR108).

File: src/supermarket/app/models/cookbook_upload/archive.rb (hunk 1 above)
Comment (robbkidd) [reply src/supermarket/app/models/cookbook_upload/archive.rb]: Added a code comment with basically those words.

**Comment by robbkidd** [comment]
I missed adding `libarchive` to the omnibus build! Working on that and a test build locally of the package.

**Comment by robbkidd** [comment]
OK. `libarchive` added to be included in omnibus build. I'm pretty sure the @chef/jex-team does not have any concerns with `libarchive` also being added to the Travis build. So ... shipping it.

**Pull request merged**